pkg/activator/net/throttler.go

```diff
@@ -386,18 +385,8 @@ func assignSlice(trackers []*podTracker, selfIndex, numActivators, cc int) []*po
 	x := append(trackers[:0:0], trackers[bi:ei]...)
 	if remnants > 0 {
 		tail := trackers[len(trackers)-remnants:]
-		// We shuffle the tail, to ensure that pods in the tail get better
-		// load distribution, since we sort the pods above, this puts more requests
-		// on the very first tail pod, than on the others.
-		rand.Shuffle(remnants, func(i, j int) {
-			tail[i], tail[j] = tail[j], tail[i]
-		})
-		// assignSlice is not called for CC=0 case.
-		dcc := int(math.Ceil(float64(cc) / float64(numActivators)))
-		// This is basically: x = append(x, trackers[len(trackers)-remnants:]...)
-		// But we need to update the capacity.
-		for _, t := range tail {
-			t.UpdateConcurrency(dcc)
```

Comment:
Can you comment on the deletions here (in GitHub) and document the code a bit.

I need to validate my understanding. The way I read the change is we are no longer distributing tail elements across all activators and instead having activators pick up an additional element

eg.
1. we have 20 pods and 3 activators -> we'd get 2 remnants so activator with index 0,1 would each pick up a unique tracker
2. we have 24 pods and 5 activators -> we'd get 4 remnants so the activator 0,1,2,3 would each pick up a unique tracker

What's the reason for dropping the `t.UpdateConcurrency(dcc)` call?  I'm a bit lost why the code prior called `UpdateConcurrency` on the tail trackers but not the picked indicies?

Reply:
> What's the reason for dropping the t.UpdateConcurrency(dcc)

So the test change where the total capacity changed from 15 to 20 highlighted that even though the tracker was shared amongst the multiple activators before - the container concurrency is reduced for each tracker.

I'm surprised this triggered LB issues because that remnant pod tracker should still have a cc=1.

I wonder if there's nothing wrong with our capacity calculation but actually the LB implementation - for cc=1 it uses `firstAvailableLBPolicy`

Reply:
Thanks for your comments!

> Can you comment on the deletions here (in GitHub) and document the code a bit.
Sure, I will add comments on the code.

> I need to validate my understanding. The way I read the change is we are no longer distributing tail elements across all activators and instead having activators pick up an additional element
Yes, I think your understanding is same as mine.

> What's the reason for dropping the t.UpdateConcurrency(dcc) call? I'm a bit lost why the code prior called UpdateConcurrency on the tail trackers but not the picked indicies?
I think the reason why the code prior called UpdateConcurrency was to ensure that the total capacity of each tail tracker across all activators matched the desired cc.
However, in this PR, each remnant tracker is now assigned to a unique activator. As a result there is no need to call UpdateCapacity anymore.


> I'm surprised this triggered LB issues because that remnant pod tracker should still have a cc=1.
> 
> I wonder if there's nothing wrong with our capacity calculation but actually the LB implementation - for cc=1 it uses firstAvailableLBPolicy
While the call of UpdateConcurrency here may not function as intended, this issue's root cause is not here. Instead, activator itself's capacity calculation is the cause.

If I have misunderstood anything or if you have further questions, please let me know.

Reply:
> If I have misunderstood anything or if you have further questions, please let me know.

Bear with me - I'm also learning the code so I'm walking through it so I understand the implication of your changes

> If podTracker=3, assignedTracker=2 and activatorCount=2, the expected targetCapacity should be 2. However, in the current implementation targetCapacity is calculated as 1.

When CC=1 the activator uses an infinite breaker which _always_ reports a capacity of 1 when you update the breaker's capacity >= 1. If the capacity is 1 then it will let in as many requests as possible

That's why I was wondering if we're still including excess destinations to all activators (and reducing the container concurrency) it seems like there's an issue with the load balancing.

With the old code I added a test to see the output in your original situation (cc=1, pods=5, activators=2)
```go
func TestBlah(t *testing.T) {
	logger := TestLogger(t)
	cc := 1
	numActivators := 2

	for i := 0; i < numActivators; i++ {
		rt := &revisionThrottler{
			logger:               logger,
			breaker:              queue.NewBreaker(testBreakerParams),
			containerConcurrency: cc,
		}
		rt.numActivators.Store(int32(numActivators))
		rt.activatorIndex.Store(int32(i))
		rt.podTrackers = makeTrackers(5, 1)
		rt.breaker = newInfiniteBreaker(logger)

		rt.updateCapacity(5)
		got := rt.breaker.Capacity()
		t.Logf("Activator breaker capacity = %d", got)

		for _, tracker := range rt.assignedTrackers {
			t.Logf("dest: %s, Capacity = %d", tracker.dest, tracker.Capacity())

		}
	}
}
```

This gives the output

```
=== RUN   TestBlah
    logger.go:130: 2023-06-09T11:28:39.139-0400 DEBUG   net/throttler.go:300    Trackers 0/2: assignment: [0 1 4]
    logger.go:130: 2023-06-09T11:28:39.139-0400 INFO    net/throttler.go:318    Set capacity to 2 (backends: 5, index: 0/2)
    throttler_test.go:245: Activator breaker capacity = 1
    throttler_test.go:248: dest: 0, Capacity = 1
    throttler_test.go:248: dest: 1, Capacity = 1
    throttler_test.go:248: dest: 4, Capacity = 1
    logger.go:130: 2023-06-09T11:28:39.140-0400 DEBUG   net/throttler.go:300    Trackers 1/2: assignment: [2 3 4]
    logger.go:130: 2023-06-09T11:28:39.140-0400 INFO    net/throttler.go:318    Set capacity to 2 (backends: 5, index: 1/2)
    throttler_test.go:245: Activator breaker capacity = 1
    throttler_test.go:248: dest: 2, Capacity = 1
    throttler_test.go:248: dest: 3, Capacity = 1
    throttler_test.go:248: dest: 4, Capacity = 1
--- PASS: TestBlah (0.00s)
```

So in theory dest 4 should be receiving requests from _both_ activators which would result in some requests being queued at the queue proxy.

Given this I wondered why are we not receiving traffic and I think it's because of the load balancing policy which just iterates over the assigned trackers and picks the first one that has capacity.

https://github.com/knative/serving/blob/5a90438bbd9d7a677310c238b388d6f9a82ffe5c/pkg/activator/net/throttler.go#L182-L185

https://github.com/knative/serving/blob/5a90438bbd9d7a677310c238b388d6f9a82ffe5c/pkg/activator/net/throttler.go#L213

Thus I don't think your change this will solve your original problem. I think if you add artificial waits to your services handling requests you would eventually see all pods getting traffic. Or set your container concurrency to 4 which changes the load balancing to round robin.

So I think there's a separate fix to the original issue and we need to think about whether changing the assignment still makes sense (I think it does - just need to think about the implications a bit)

Reply:
Thank you for your comment! I got what you meant.
Your opinion is different from my understanding. I will explain it.

> When CC=1 the activator uses an infinite breaker which always reports a capacity of 1 when you update the breaker's capacity >= 1. If the capacity is 1 then it will let in as many requests as possible

This is not correct.
When CC >= 1, the infinite breaker is not used. Instead, queue.NewBreaker is called.
https://github.com/knative/serving/blob/main/pkg/activator/net/throttler.go#L179-L190

Let's consider the scenario where there are multiple activators and CC=1.
In the case of CC = 1, the capacity of each pod becomes 1.
There is no mechanism for activators to share which pods they are sending requests to.
If both activators are assigned to all pods, there is a possibility that the total requests from all activators will exceed CC. Therefore, each activator is assigned to unique pods.

Reply:
@dprotaso 

> Thus I don't think your change this will solve your original problem. I think if you add artificial waits to your services handling requests you would eventually see all pods getting traffic. Or set your container concurrency to 4 which changes the load balancing to round robin.
> 
> So I think there's a separate fix to the original issue and we need to think about whether changing the assignment still makes sense (I think it does - just need to think about the implications a bit)

As you pointed out, the issue can also be resolved by changing the load balancing mechanism to round-robin.
However, in my opinion, it would be better to modify the capacity of activator itself. The problem of this issue is not load balancing logic, but capacity of activators.

In the case of cc=1, pods=5, and activators=2:
Each activator has a capacity of 2.
The assignment of activators is as follows: activator1: pod0, pod1, pod4; activator2: pod2, pod3, pod4.

When the hey command is executed:
`hey -t 0 -n 10000 -c 100 <address>`
The activator is designed not to process requests beyond its capacity.
If a request is sent to activator1, it first tries to send the request to pod0. If pod0 has available capacity, the request is sent there. If pod0 doesn't have available capacity, the request is sent to pod1.
So what happens if neither pod0 nor pod1 can receive the request?
In this case, the activator has reached its capacity limit of 2, so it won't attempt to send the request to the pod in the first place.
When there is available capacity in pod0 or pod1 again, it will attempt to send the request to the pod. That's why pod4 wouldn't get any traffic from activator1 and activator2 (you can observe it by reproducing the issue).

Changing to round-robin will likely solve this bug as well. However, even with round-robin, the issue of activator capacity remains unresolved. Therefore, I believe it would be better to fix the activator's capacity itself, which is the modification in my PR.
Also, if we change to round-robin even when cc <= 3, there is no mechanism for multiple activators to share which pods they are sending requests to, so it will result in sending requests exceeding the capacity to pod4.

Sorry for long explanation.
If I have misunderstood anything or if you have further questions, please let me know.